### ISSUE TITLE
Delta backup fails with too many open files error

### ISSUE THREAD

**Issue by pawanrawal** [wal-g/wal-g#1156]
### Database name
Postgres

# Issue description

We have a large cluster with many nodes that we backup daily. Delta backups on one of the nodes fail even though base backups succeed. 

### Describe your problem
We are not able to take delta backups and are also not able to diagnose what is the problem here.

### Please provide steps to reproduce
Run `wal-g backup push` to take delta backup using `wal-g` version `1.1`

### Please add config and wal-g stdout/stderr logs for debug purpose

also you can use WALG_LOG_LEVEL=DEVEL for logs collecting

<details><summary>If you can, provide logs</summary>
<p>

INFO: 2021/11/22 00:00:02.533401 Selecting the latest backup as the base for the current delta backup...
INFO: 2021/11/22 00:00:19.568035 Delta backup from base_0000000100011B4E0000006A with LSN 11b4e6a5f4800.
INFO: 2021/11/22 00:00:19.624799 Calling pg_start_backup()
INFO: 2021/11/22 00:01:31.357410 Delta backup enabled
INFO: 2021/11/22 00:01:31.357478 Starting a new tar bundle
INFO: 2021/11/22 00:01:31.357551 Walking ...
INFO: 2021/11/22 00:01:31.358115 Starting part 1 ...
INFO: 2021/11/22 00:01:31.366777 Starting part 2 ...
INFO: 2021/11/22 00:01:31.368371 Starting part 3 ...
...
...
WARNING: 2021/11/22 00:03:03.791765 failed to read file '/base/16384/760343710' as incremented
WARNING: 2021/11/22 00:03:04.383357 failed to read file '/base/16384/782765281' as incremented
WARNING: 2021/11/22 00:03:04.597364 failed to read file '/base/16384/814418038' as incremented
WARNING: 2021/11/22 00:03:04.701620 failed to read file '/base/16384/82914570' as incremented
WARNING: 2021/11/22 00:03:05.349695 failed to read file '/base/16384/909803074' as incremented
WARNING: 2021/11/22 00:03:05.536041 failed to read file '/base/16384/924390002' as incremented
WARNING: 2021/11/22 00:03:05.610798 failed to read file '/base/16384/925591530' as incremented
WARNING: 2021/11/22 00:03:05.615053 failed to read file '/base/16384/925603995' as incremented
WARNING: 2021/11/22 00:03:05.628835 failed to read file '/base/16384/925644479' as incremented
WARNING: 2021/11/22 00:03:05.630097 failed to read file '/base/16384/925647963' as incremented
WARNING: 2021/11/22 00:03:05.714253 failed to read file '/base/16384/926142825' as incremented
WARNING: 2021/11/22 00:03:06.061318 failed to read file '/base/16384/946933811' as incremented
WARNING: 2021/11/22 00:03:06.063391 failed to read file '/base/16384/946947060' as incremented
WARNING: 2021/11/22 00:03:06.067203 failed to read file '/base/16384/947213781' as incremented
WARNING: 2021/11/22 00:03:06.078828 failed to read file '/base/16384/948702325' as incremented
INFO: 2021/11/22 00:03:06.889111 Finished writing part 43.
INFO: 2021/11/22 00:03:08.093928 Finished writing part 14.
INFO: 2021/11/22 00:03:12.779734 Finished writing part 17.
INFO: 2021/11/22 00:03:14.761588 Finished writing part 46.
INFO: 2021/11/22 00:03:15.757977 Finished writing part 44.
ERROR: 2021/11/22 00:03:15.758083 HandleWalkedFSObject: handle tar failed: startReadingFile: failed to open file '/data/12/base/16384/1024282376'
: open /data/12/base/16384/1024282376: too many open files

</p>
</details>

**Comment by usernamedt** [comment]
Hi! What is your opened file descriptors limit?

**Comment by pawanrawal** [comment]
> Hi! What is your opened file descriptors limit?

It's 65536. Interestingly I have the same limit on the rest of the servers where I run wal-g but none of them faces this issue.  `WALG_UPLOAD_CONCURRENCY` is set to 72 so I reckon `wal-g` should only be opening 72 files at a time?

**Comment by usernamedt** [comment]
> > Hi! What is your opened file descriptors limit?
> 
> It's 65536. Interestingly I have the same limit on the rest of the servers where I run wal-g but none of them faces this issue. `WALG_UPLOAD_CONCURRENCY` is set to 72 so I reckon `wal-g` should only be opening 72 files at a time? I am wondering if there is a bug somewhere and `wal-g` ends up opening a larger number of files when it can't read the file as incremented?

Actually, the `WALG_UPLOAD_CONCURRENCY` affects only the `wal-push` command. To control the number of the parallel tarballs being written, use the `WALG_UPLOAD_DISK_CONCURRENCY` setting. It would be great to get the output of the `lsof -p *WAL-G PID*` command during the backup-push run on the problematic server.

**Comment by pawanrawal** [comment]
I did check the output of `lsof -p *WAL-G PID*` in a loop, below is the output. At any given time the number of open file descriptors can even touch 4000. I had to increase the limit for the number of open file descriptors to 4000 to be able to get it to work as with the default limit of 1024 the `wal-g` process crashes. 

```
while true; do ls /proc/1424557/fd/ | wc -l; sleep 1; done;
```

Output
```
114
128
208
245
373
684
973
1054
1113
1134
1145
1172
1229
1391
1612
1710
1718
1879
1890
2000
2186
2271
2296
2524
2846
2854
2852
2895
2991
2998
3015
3016
3015
3015
3016
3017
3020
3019
2915
2651
```

I don't have the time to delve deeper into this right now but seems like the limit on the number of open files isn't being enforced properly or the files aren't being closed properly by `wal-g`.

**Comment by pawanrawal** [comment]
@usernamedt any ideas as to what is the issue here?

**Comment by usernamedt** [comment]
> @usernamedt any ideas as to what is the issue here?

Hi! Does the issue persist with the file descriptors limit of 4000?

**Comment by pawanrawal** [comment]
I have investigated this some more and feel like there is a bug in `wal-g` here. So I stopped the Postgres instance and enabled checksums on it using [pg_checksums](https://www.postgresql.org/docs/14/app-pgchecksums.html). `pg_checksums` didn't show any invalid blocks when I ran the verify checksums command using it. I specifically also checked the files that are part of the WARNING output log to verify them but I could find any bad blocks. Here is an example output.

```
root@ec2-instance:/usr/lib/postgresql/12/bin# ./pg_checksums -D /data/12 --filenode=100771765
Checksum operation completed
Files scanned:  1
Blocks scanned: 169
Bad checksums:  0
Data checksum version: 1
```

Enabling the debug output on `wal-g` I see this debug log
```
DEBUG: 2021/12/12 23:48:35.042977 Invalid page  1  page header  &{0 0 53723 0 840 1664 8176 8196}
WARNING: 2021/12/12 23:48:35.043001 failed to read file '/base/16384/100771765' as incremented
```

This tells me that `pg_checksum` considers the file to be healthy but `wal-g` somehow thinks that the page header is invalid. When `wal-g` encounters a lot of such files, the backup can't complete because it ends up opening too many files. So I can see two bugs here with `wal-g`. One is that it doesn't respect or has any reasonable limit to the number of open files and the second is that the logic for what constitutes a valid page header seems broken.

**Comment by usernamedt** [comment]
Hi!
```
WARNING: 2021/12/12 23:48:35.043001 failed to read file '/base/16384/100771765' as incremented
```
AFAIK this warning might occur when reading the torn page. This is a normal situation so it can be safely ignored. 

Indeed, it looks like there is a bug in the [ReadIncrementalFile](https://github.com/wal-g/wal-g/blob/0261341f03bafb48960dbb5835d2b921cbb92b14/internal/databases/postgres/pagefile.go#L134) function because it does not close the opened file on encountering such errors.

**Comment by serprex** [comment]
@pawanrawal does latest fix your issue?

**Comment by pawanrawal** [comment]
Hey @serprex @usernamedt, I tried out the binary from master and seems like the problem has been fixed. Thanks for looking into this. Would the fix be backported to the `v0.2.x` releases?

**Comment by serprex** [comment]
No commits have been made to `0.2.x` tags since May. `1.0` was since released. So I'm guessing no
